src/core/sequence.ml

```diff
@@ -0,0 +1,41 @@
+
+type +'a node =
+   | Nil
+   | Cons of 'a * 'a t
+and 'a t = unit -> 'a node
+
+let empty () = Nil
+
+let return x () = Cons(x, empty)
+
+let rec map f i () = match i () with 
```

Comment:
First, why `match ... with` is bad here? I understand you mentioned it a couple of times before but I failed to grasp why this is a bad pattern. Second, I tried a bit, and found this one really hard to be rewritten using only `function` since we have to match against `i ()`.